File: python/queen-attack/queen_attack.py
"""Queen Attack."""
from __future__ import annotations

class Queen:
    """A cartesian coordinate."""

    def __init__(self, row: int, column: int):
        """Initialize."""
        for name, val in [("row", row), ("column", column)]:
            if val < 0:
                raise ValueError(f"{name} not positive")
            if val >= 8:
                raise ValueError(f"{name} not on board")
        self.position = (row, column)

    def can_attack(self, other: Queen) -> bool:
        """Return if two queens can attack."""
        if self.position == other.position:
            raise ValueError("Invalid queen position: both queens in the same square")
        delta = {abs(self.position[i] - other.position[i]) for i in (0, 1)}
        # `0 in delta` indicates the queens are on the same row or column.
        # `len(delta) == 1` indicates the x and y delta are the same, ie on a diagonal.
        return 0 in delta or len(delta) == 1

    def print(self):
        """Print the coordinate."""
        print(self.position)

File: go/queen-attack/queen_attack.go
package queenattack

import "fmt"

func position(p string) (int, int, error) {
	r := []rune(p)
	if len(r) != 2 {
		return 0, 0, fmt.Errorf("invalid position, %s", p)
	}
	if r[0] < 'a' || r[0] > 'h' {
		return 0, 0, fmt.Errorf("invalid file, %c", r[0])
	}
	if r[1] < '1' || r[1] > '8' {
		return 0, 0, fmt.Errorf("invalid file, %c", r[1])
	}
	return int(r[0] - 'a'), int(r[1] - '1'), nil
}

// CanQueenAttack returns if a queen can attack another.
func CanQueenAttack(whitePosition, blackPosition string) (bool, error) {
	if whitePosition == blackPosition {
		return false, fmt.Errorf("positions cannot be the same, %s", whitePosition)
	}
	wX, wY, err := position(whitePosition)
	if err != nil {
		return false, err
	}
	bX, bY, err := position(blackPosition)
	if err != nil {
		return false, err
	}

	canAttack := wX == bX || wY == bY || (wY-bY == wX-bX) || (wY-bY == bX-wX)
	return canAttack, nil
}
